### ISSUE TITLE
Nested classes/interfaces cannot be resolved in child classes

### ISSUE THREAD

**Issue by roymengeo** [redhat-developer/vscode-java#3076]
[provide a description of the issue]

##### Environment
- Operating System: Windows 10
- JDK version: Java 17
- Visual Studio Code version: 1.77.3
- Java extension version: 1.17

##### Steps To Reproduce
See below.

##### Current Result
Starting with version v1.15 of the extension, nested classes/interfaces cannot be resolved in child classes any more. Assume we have the following situation:

class A {
   ...
   protected class Inner {
       ...
   }
}

class B extends A {
   ...
  public void foo(Inner bar) {
   ...
  }
}

In this case, class B will not compile but result in the following error: "Inner cannot be resolved to a type". In order to make it compile I need to replace Inner by A.Inner like so:

class B extends A {
   ...
  public void foo(A.Inner bar) {
   ...
  }
}

##### Expected Result

Using Inner within B without prefixing its name by A used to work in all extension version < 1.14. It also works when using a current version of the Eclipse IDE or when compiling the code with Maven. It seems to me that the compiler used by VSCode is a little to strict here.

**Comment by fbricon** [comment]
I can't reproduce, with 1.17 or latest 1.18 pre-release. Please provide a sample project reproducing this issue.

**Comment by roymengeo** [comment]
Thanks for your quick response. I just tried to create a minimal example reproducing the problem but I failed - the problem did not show up. It is only there in the large code base of an app I am working on. I will give it another try in a few days.

**Comment by pioneer2k** [comment]
I have the same issue in my project since version 1.15. With 1.14 everything is fine.
We are using Vaadin UI Framework. Some of our classes extending some Vaadin classes like VerticalLayout.
Example: 
public class WSRequestView extends VerticalLayout

The VerticalLayout implements an interface that has a public enum, that should be visible for all child classes:
public interface FlexComponent extends HasOrderedComponents, HasStyle, HasSize {
    public enum Alignment {
    
When we use the enum Alignment like in this example:
		final var buttonLayout = new HorizontalLayout();
		buttonLayout.setDefaultVerticalComponentAlignment(Alignment.END);
then in version 1.14 everything is fine, but version 1.15+ complains about the enum Alignment and says that the class in unknown. If I add an additional import for that enum it is ok again, but VSCode gives me a hint to remove the unneccesary import.

**Comment by fbricon** [comment]
Please, provide a sample project

**Comment by pioneer2k** [comment]
I have the same problem as mentioned above. On small example projects the problem is not there.....

**Comment by drub0y** [comment]
TL;DR - I think something is going on in the language server type resolution where, when a nested type name conflicts with other types from other packages, it just can't figure out the correct resolution.

I'm on `1.18.0` and I'm having this problem. I tried to create a simple repro, but couldn't get it to happen right away. Then I realized that the nested type, which happens to be called `Expiry`, conflicts the `Expiry` type from caffeine which also happens to be a dependency of the project where I'm experiencing this. However, it is not like the package is imported so it shouldn't be causing a name collision which is why there are no problems at compile time, however the language server says the type cannot be resolved. Additionally, it suggests importing the base classes nested `Expiry` and that effectively "no-ops" when you click it.

**Comment by fbricon** [comment]
@drub0y so you do have a sample project ready to be shared right?

**Comment by drub0y** [comment]
> @drub0y so you do have a sample project ready to be shared right?

I honestly can not get it to repro in a clean project. I built out the exact same package structure and class hierarchy with a nested enum named `Expiry` with caffeine in the project and can't get it to trigger the same thing.

**Comment by drub0y** [comment]
I got it, it has to do with referring to the class in an annotation on the class itself. Simplest repo attached.

[nested-type-error-repro.gz](https://github.com/redhat-developer/vscode-java/files/11445271/nested-type-error-repro.gz)

**Comment by fbricon** [comment]
@drub0y thanks for the test project. So it looks like a regression in upstream eclipse JDT, although I'm not able to reproduce in Eclipse itself, but older vscode-java version (like v1.14) seem to work.
@snjeza can you please take a look?

**Comment by fbricon** [comment]
@roymengeo  nevermind, I forgot I already asked ;-)

**Comment by snjeza** [comment]
I can't reproduce the issue using VS Code 1.18 and pre release.
A related issue - https://github.com/eclipse-jdt/eclipse.jdt.core/issues/612

**Comment by snjeza** [comment]
> I got it, it has to do with referring to the class in an annotation on the class itself. Simplest repo attached.

@drub0y @fbricon This issue can be reproduced in VS Code.
It can't be reproduced in Eclipse.

**Comment by roymengeo** [comment]
Just adding a comment here to say:

1. Thanks guys for investigating this!
2. Snjeza modified his last comment after initially posting it. The email notification sent out by GitHub did only reflect the first version of this comment where he could not reproduce the issue. So you might have missed that afterwards he was able to nail it as an "upstream Eclipse issue".

Thanks again,
Roy

**Comment by fbricon** [comment]
@roymengeo Snjeza's a she though :-)

**Comment by roymengeo** [comment]
Whoops, I am sorry!

**Comment by drub0y** [comment]
> I can't reproduce your issue but can reproduce @drub0y issue https://github.com/redhat-developer/vscode-java/issues/3076#issuecomment-1542691617 with `"java.compile.nullAnalysis.mode": "enabled"`

I can absolutely confirm that I definitely have that setting enabled. Disabling makes the problem go away immediately and, surprise, re-enabling it instantly brings it back.

**Comment by rgrunber** [comment]
Based on the referenced commits, as soon as we update to something like [`I20230702-1800`](https://github.com/eclipse-jdt/eclipse.jdt.core/releases/tag/I20230702-1800) in JDT-LS, the problem should go away.
